### DISCUSSION TITLE
How do I get a static list of all supported protocols?

### DISCUSSION THREAD

**Discussion by nathanielc** [libp2p/rust-libp2p#4982] [Q&A]
With the changes in 0.52 its now possible to dynamically get a list of supported protocols. However I am interested (for diagnostic reasons) in getting the union of supported protocols of the local node regardless of if there is an active connection using that protocol. 

As far as I can tell the only way to get a list of protocols now is to listen for `ConnectionEvent::LocalProtocolsChange` events. However this implies that whatever protocol is listening for those events is shared between all connected peers. Otherwise the handler does not get the `ConnectionEvent::LocalProtocolsChange` events themselves. Do I miss understand?

Is there a way to directly ask a behavior for its static set of protocols? Or maybe a way for the `identify` behavior to expose the union of all locally supported protocols across all connections? 

Preferably I can get this set before even a single connection is made. Thoughts?

**Comment by mxinden** [comment]
> As far as I can tell the only way to get a list of protocols now is to listen for `ConnectionEvent::LocalProtocolsChange` events. However this implies that whatever protocol is listening for those events is shared between all connected peers. Otherwise the handler does not get the `ConnectionEvent::LocalProtocolsChange` events themselves. Do I miss understand?

I am not sure this is correct. `LocalProtocolsChange` is scoped to a connection, i.e. it implies that the local node supports a protocol on the specific connection.

> Is there a way to directly ask a behavior for its static set of protocols? Or maybe a way for the `identify` behavior to expose the union of all locally supported protocols across all connections?

Note that it is not the `NetworkBehaviour`, but the `ConnectionHandler`. In other words, it is not the main state machine, but the connection state machine.

> Preferably I can get this set before even a single connection is made. Thoughts?

Today, I don't see a way other than implementing your own `ConnectionHandler` that listens for the local protocol change events only.

**Reply by nathanielc** [reply]
>Note that it is not the NetworkBehaviour, but the ConnectionHandler. In other words, it is not the main state machine, but the connection state machine.

Yeah that makes sense. The reason I asked about it from the Behaviour context is that the Behaviour knows the static set of possible protocols because the behaviour handles calling `listen_protocol` on each of the connection handlers. I guess the challenge here is the behaviours can dynamically create connection handlers and so its not well defined statically. 

> Today, I don't see a way other than implementing your own ConnectionHandler that listens for the local protocol change events only.

I tried this, but my testing showed that I didn't get the ConnectionHandler::LocalProtocolsChange event. My assumption is that because my new connection handler was for a protocol that the remote node did not support.



FWIW for now i have hard coded my list of local protocols for diagnostic reasons. I may not even need this in the future. Thanks for confirming that I didn't miss something obvious.

**Reply by thomaseizinger** [reply]
> > Note that it is not the NetworkBehaviour, but the ConnectionHandler. In other words, it is not the main state machine, but the connection state machine.
> 
> Yeah that makes sense. The reason I asked about it from the Behaviour context is that the Behaviour knows the static set of possible protocols because the behaviour handles calling `listen_protocol` on each of the connection handlers.

It seems there is a misunderstanding here. `NetworkBehaviour`s and `ConnectionHandler`s run in separate tasks and they communicate via message-passing. Thus the `NetworkBehaviour` does **not** call `listen_protocol`.

- `NetworkBehaviour`s are composed into a tree (same as deriving `serde::Serialize` just applies recursively)
- `ConnectionHandler`s are also composed into a tree
- 1 `ConnectionHandler` is instantiated per _physical_ connection (i.e. TCP, QUIC, etc). The `NetworkBehaviour` is tasked with this via `handle_{in,out}bound_....`
- Once constructed, a new task is spawned into your configured runtime that bridges between the established connection (i.e. `StreamMuxer` and the `ConnectionHandler` state machine).

Because a `NetworkBehaviour` gets to chose on a _per connection_ basis, which `ConnectionHandler` it returns, there is no such thing as a "static list of supported protocols". At least not generally. It may be that the set of `NetworkBehaviour`s that you use always supports all protocols on all connections but this is not universally true. For example, `libp2p::relay` will refuse to establish further relayed connections over an existing relayed connection. This is a decision we make on a per-connection basis.

> > Today, I don't see a way other than implementing your own ConnectionHandler that listens for the local protocol change events only.
> 
> I tried this, but my testing showed that I didn't get the ConnectionHandler::LocalProtocolsChange event.

You will receive this event as the very first thing once the connection is established.

> My assumption is that because my new connection handler was for a protocol that the remote node did not support.

This assumption is not correct. Unless you have a protocol such as "identify" in place, we have no idea about learning which protocols the remote supports. Certainly, we cannot know this prior to such a protocol executing.

> FWIW for now i have hard coded my list of local protocols for diagnostic reasons. I may not even need this in the future. Thanks for confirming that I didn't miss something obvious.

In general, it is best if you don't program against the assumption of a statically supported list of protocols. Each protocol should fail gracefully if the remote doesn't support it. In other words, handle `StreamUpgradeError::NegotiationFailed` properly. For inbound streams, there is (somewhat obviously) nothing you need to do.

**Reply by nathanielc** [reply]
Thanks for the detailed explanations. 

Yeah I understand the separate tasks and that the Behaviour itself is not calling listen_protocol, I should have been more specific, apologies for the confusion. 

And generally we are not programing against a static set of protocols, rather just wanted to expose the set for diagnostics. However its clear that's not a good approach and will likely change the upstream logic accordingly.



I am however still confused about this bit:


> > > Today, I don't see a way other than implementing your own ConnectionHandler that listens for the local protocol change events only.
> > 
> > 
> > I tried this, but my testing showed that I didn't get the ConnectionHandler::LocalProtocolsChange event.
> 
> You will receive this event as the very first thing once the connection is established.
> 
> > My assumption is that because my new connection handler was for a protocol that the remote node did not support.
> 
> This assumption is not correct. Unless you have a protocol such as "identify" in place, we have no idea about learning which protocols the remote supports. Certainly, we cannot know this prior to such a protocol executing.


I had assumed that my connection handler would not receive ConnectionHandler events if it didn't upgrade the connection for the specific protocol of the handler. Do all connection handlers for a connection receive all events even if the connection is not fully negotiated ? 

For example if I have two Behaviours  `identify` and a custom one `foo`. Assuming all nodes I connect to support `identify` and only some support `foo`. Will the local `foo` connection handlers to remote nodes that do not support `foo` continue to receive `ConnectionHandler::LocalProtocolsChange` events?

**Reply by thomaseizinger** [reply]
> For example if I have two Behaviours `identify` and a custom one `foo`. Assuming all nodes I connect to support `identify` and only some support `foo`. Will the local `foo` connection handlers to remote nodes that do not support `foo` continue to receive `ConnectionHandler::LocalProtocolsChange` events?

All `ConnectionHandler`s are composed into a tree using `ConnectionHandlerSelect`:

- `ConnectionHandlerSelect`
	- `identify::Handler`
	- `foo::Handler`

Inbound & outbound streams and events from the behaviour are only delivered to **one** of the subhandlers. All other events are delivered to all sub handlers, see here:

https://github.com/libp2p/rust-libp2p/blob/577e387b4cdad281d069ee6cd532ba0f4241953d/swarm/src/handler/select.rs#L274-L349

**Reply by nathanielc** [reply]
Thanks!
